[docs][IRPGO]Document two binary formats for instrumentation-based profiles, with a focus on IRPGO.

mingmingl-llvm opened the pull request:
Github review tool renders the rich diff ([example](https://github.com/llvm/llvm-project/pull/76105/files?short_path=b1805ae#diff-b1805ae3bd5b5cf0c69249b7df329ef8dd7cbfe322ae004c61f4bd507d2a87e6))

mingmingl-llvm committed:
[compiler-rt][test]Mark thinlto icp test as UNSUPPORTED. Test failed when building instrumented binary

mingmingl-llvm committed:
Revert "[compiler-rt][test]Mark thinlto icp test as UNSUPPORTED. Test failed when building instrumented binary"

This reverts commit e3bfecf57ff7440ed50b51114df0d726c0e16b61.

mingmingl-llvm committed:
[docs][IRPGO]Document two binary formats for IRPGO profiles

minglotus-6 commented:
Please kindly add who might be interested in this, thanks!

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+Overview
+==========
+
+IR-based instrumentation (IRPGO) and its context-sensitive variant (CS-IRPGO)
```
Instrumentation PGO (both IR based and Frontend based).

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+IR-based instrumentation (IRPGO) and its context-sensitive variant (CS-IRPGO)
+inserts `llvm.instrprof.*` `code generator intrinsics <https://llvm.org/docs/LangRef.html#code-generator-intrinsics>`_
+in LLVM IR to generate profiles. This document describes two binary profile
+formats (raw and indexed) used by IR-based instrumentation.
```
by instrumentation.

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+
+.. note::
+
+  Both the compiler-rt profiling infrastructure and profile format are general
```
Coverage test uses (frontend) PGO instrumentation and coverage mapping.  The format for coverageMap is not included in this document. Similarly the temporal profiling is not covered here.

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+The raw PGO profile is generated by running the instrumented binary. It is a
+memory dump of the profile data.
+
+Two kinds of frequently used profile information are function's basic block
```
The instrumented binary currently collects two kinds of profile data: ...

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+memory dump of the profile data.
+
+Two kinds of frequently used profile information are function's basic block
+counters and its (various flavors of) value profiles. A function's profiled
```
The profile data for a function can span ..

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+General Storage Layout
+-----------------------
+
+A raw profile for an executable [1]_ consists of a profile header and several
```
also shared libary.

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+-----------------------
+
+A raw profile for an executable [1]_ consists of a profile header and several
+sections. The storage layout is illustrated below. Generally, when raw profile
```
when the raw profile ..

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+
+``Magic``
+  With the magic number, data consumer could detect profile format and
+  endianness of the data, and quickly tells whether/how to continue reading.
```
remove 'quickly'

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,387 @@
+  referenced as explained by calculation-of-counter-offset_ .
+
+``NamesDelta``
+  Records the in-memory address of compressed name section. Not used except for
```
May be uncompressed too.

mingmingl-llvm committed:
resolve feedback

mingmingl-llvm reviewed (commented):
The file name should probably be updated to `InstrumentationPGOProfileFormat` . Will wait and do this later as one-off to minimize diff..

MaskRay commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,395 @@
+===================================
```
Perhaps name this ProfileFormat.rst, since so much is shared with pure code coverage applications. It's fine that the doc currently focuses on PGO.

minglotus-6 replied:
> Perhaps name this ProfileFormat.rst, since so much is shared with pure code coverage applications.

Ack. I wonder if we want to use `InstrumentationProfileFormat.rst` since SamplePGO uses different format.

snehasish commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,395 @@
+
+::
+
+                            +-----------------------+---+
```
Should we also add a comment in the code to update this documentation when the format changes?

snehasish commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,395 @@
+  identical functions.
+
+``FuncHash``
+  A fingerprint of the function's control flow graph.
```
This includes CFG plus some more stuff (memory ops I think). Can you put in a link to the code?

snehasish commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,395 @@
+
+Bitmap
+^^^^^^^
+This section is used for source-based MC/DC code coverage. Check out `Bitmap RFC`_
```
Expand MC/DC?

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,395 @@
+
+Function PGO data
+^^^^^^^^^^^^^^^^^^
+This section stores functions and their PGO profiling data as an on-disk hash
```
Profile data for functions with the same name are grouped together and share one hash table entry (the functions may come from different shared libraries for instance). The profile data for them are organized as a sequence of key-value pair where the key is the funcHash (CFG based for IR PGO), and the value is profile counters for the function.

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,395 @@
+``FuncHash``
+  A fingerprint of the function's control flow graph.
+
+``CounterPtr``
```
The description is correct.  (My comment was based on old implementation before recent changes).

ZequanWu replied:
`CounterPtr` is still a relative address (`__profc_foo` - `__profd_foo`) in default mode. Under binary profile correlation mode, it will just be the absolute address of the counter `__profc_foo`.

david-xl commented on llvm/docs/PGOProfileFormat.rst:
```diff
@@ -0,0 +1,395 @@
+
+* The profile header records `CounterDelta` with the value as `start(__llvm_prf_cnts) - start(__llvm_prf_data)`.
+  We will call it `CounterDeltaInitVal` below for convenience.
+* For each profile data record, `CounterPtrN` is recorded as `start(Counter) - start(ProfileData)`.
```
discard this comment.

mingmingl-llvm committed:
incorporate feedbacks
* Some rewording as suggseted.
* Add link to code at a specific commit at a few places.
* Mention it explicitly when fields might have a different semantics in
  non-IRPGO case.
* Mention version compatibility guarantees explicitly for both formats,
  and add more details on endianness handling for raw profiles.
* Add code comment to ask for doc update if appropriate.

mingmingl-llvm committed:
rename file to InstrProfileFormat.rst

mingmingl-llvm committed:
correct footnote numbering

mingmingl-llvm committed:
one-liner fix for one footnote

ZequanWu commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,457 @@
+  The number of bytes in the name section.
+
+``CountersDelta``
+  In the IRPGO case [4]_, this field records the in-memory address difference
```
I think "In the IRPGO case [4]_," can be removed, since `CountersDelta` and `BitmapDelta` are in-memory address differences in both IRPGO and Frontend instrumentations. And they are 0 only under lightweight instrumentations (debug-info correlation or binary correlation). (BTW, the current lightweight correlation mode doesn't set `BitmapDelta` to 0 which is a bug)

minglotus-6 replied:
sounds good. I reworded this. PTAL.

ZequanWu commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,457 @@
+    example, for `binary profile correlation`_, it represents the counter address.
+    When in doubt, check source code.
+
+``BitmapPtr``
```
Similar to CounterPtr, BitmapPtr is also an absolute address of the symbol under binary correlation.

minglotus-6 replied:
I update the description of the `CounterPtr` and `BitmapPtr`. PTAL.

mingmingl-llvm committed:
incorporate review feedback

mingmingl-llvm committed:
add two arrows for ascii graph

mingmingl-llvm committed:
resolve review feedback

minglotus-6 commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,473 @@
+Some fields are used to reference data from other sections in the profile.
+The fields are documented as follows:
+
+.. _`__llvm_profile_data`: https://github.com/llvm/llvm-project/blob/7c3b67d2038cfb48a80299089f6a1308eee1df7f/compiler-rt/lib/profile/InstrProfiling.h#L25
```
updated the link,  thanks!

minglotus-6 commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,473 @@
+        + --> start(__llvm_prf_cnts) --> +---------------------+       ||    |
+                                         |        ...          |       ||    |
+                                         +---------------------+  -----||----+
+                                         |      Counter 1      |       ||
```
done.

david-xl commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,476 @@
+
+* The profile header records `CounterDelta` with the value as `start(__llvm_prf_cnts) - start(__llvm_prf_data)`.
+  We will call it `CounterDeltaInitVal` below for convenience.
+* For each profile data record, `CounterPtrN` is recorded as `start(Counter) - start(ProfileData)`.
```
Counter --> CounterN, ProfileData --> ProfileDataN. 

Also describe that DataN is the N th entry in __llvm_prf_data, and CounterN is the corresponding profile counters.

minglotus-6 replied:
done.

mingmingl-llvm committed:
resolve review feedback

minglotus-6 commented:
thanks for all the reviews! I'll wait a little bit more for review comments, and plan to submit it in early Wednesday if the pull request looks good.

p.s. I'm likely late in the party..  just realized Github UI displays rich diff (like https://github.com/llvm/llvm-project/pull/76105/files?short_path=b1805ae#diff-b1805ae3bd5b5cf0c69249b7df329ef8dd7cbfe322ae004c61f4bd507d2a87e6) with a click for review, which means the [workaround](https://github.com/llvm/llvm-project/pull/76105#issue-2051391300) is not necessary.

snehasish commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,478 @@
+
+* The profile header records `CounterDelta` with the value as `start(__llvm_prf_cnts) - start(__llvm_prf_data)`.
+  We will call it `CounterDeltaInitVal` below for convenience.
+* For each profile data record `ProileDataN`, `CounterPtr` is recorded as `start(CounterN) - start(ProfileDataN)`,
```
typo: ProfileDataN

minglotus-6 replied:
thanks for the catch. Fixed it.

snehasish commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,478 @@
+===========================
+
+Indexed profiles are generated from `llvm-profdata`. In the indexed profiles,
+function PGO data are organized as on-disk hash table such that compilers could
```
nit: s/could/can

minglotus-6 replied:
done.

snehasish commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,478 @@
+function PGO data are organized as on-disk hash table such that compilers could
+look up PGO data for functions in an IR module.
+
+Compilers and tools must retain backward compatibility with indexed PGO profiles.
```
Spell it out? i.e. "older profiles must be readable by newer tools" or something like that...

minglotus-6 replied:
done.

snehasish commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,478 @@
+
+Binary Ids
+^^^^^^^^^^^^^^^^^^^^^^
+The section is used to carry on binary-id information from raw profiles.
```
Add a link to the RFC: https://discourse.llvm.org/t/rfc-adding-binary-id-into-llvm-profiles/58465

minglotus-6 replied:
done.

snehasish reviewed (approved):
lgtm with some minor nits.

mingmingl-llvm committed:
resolve review feedback, with two minor updates
1. Use <mangled-name> in IRPGO name format given the recent fix in pull
   request 76994.
2. In UserGuides.html, use 'instrumentation-based profiles' (not IRPGO
   profiles) to keep consistent with filename.

MaskRay commented on compiler-rt/include/profile/InstrProfData.inc:
```diff
@@ -123,6 +123,8 @@ INSTR_PROF_VALUE_NODE(PtrToNodeT, llvm::PointerType::getUnqual(Ctx), Next, \
 
 /* INSTR_PROF_RAW_HEADER  start */
 /* Definition of member fields of the raw profile header data structure. */
+/* Please update https://llvm.org/docs/InstrProfileFormat.html as appropriate
```
`llvm/docs/InstrProfileFormat.rst`

MaskRay commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+
+.. note::
+  Frontend-generated profiles are used together with coverage mapping for
+  `source based code coverage`_. The `coverage mapping format`_ is different from
```
source-based

MaskRay commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+``Magic``
+  Magic number encodes profile format (raw, indexed or text). For the raw format,
+  the magic number also encodes the endianness (big or little) and C pointer
+  byte size (32 or 64) of the platform on which the profile is generated.
```
"C pointer byte size (32 or 42)" can be misleading, 32/64 bytes?

Consider using C pointer size or just pointer size.

... encodes the endianness (big or little) and the pointer size (4 or 8 bytes).

MaskRay commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+  a platform with the opposite endianness and/or the other C pointer byte size.
+
+``Version``
+  The lower 32 bits specifies the actual version and the most significant 32
```
specify

MaskRay commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+
+.. _`advances`: https://github.com/llvm/llvm-project/blob/7e15fa9161eda7497a5d6abf0d951a1d12d86550/llvm/include/llvm/ProfileData/InstrProfReader.h#L456-L457
+
+Indexed PGO Profile Format
```
Indexed Profile Format

Since this is shared with coverage, omitting "PGO" seems clearer.

MaskRay commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+            |   |   |    |  +-----------------------+---+
+            |   |   |    |  |   Profile Summary     |   |
+            |   |   |    |  +-----------------------+   P
+            |   |   +------>|  Function PGO data    |   A
```
function data

(omit "PGO")

MaskRay commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+summary. For context-sensitive IR-based instrumentation PGO, this section stores
+an additional profile summary corresponding to the context-sensitive profiles.
+
+Function PGO data
```
Function Data

This is shared with coverage, so we probably want to de-emphasize "PGO".

MaskRay commented on llvm/docs/UserGuides.rst:
```diff
@@ -58,6 +58,7 @@ intermediate LLVM representation.
    NVPTXUsage
    Phabricator
    Passes
+   InstrProfileFormat
```
This is alphabetically ordered.

minglotus-6 commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+
+``CountersDelta``
+  This field records the in-memory address difference between the `profile metadata`_
+  and counter section in the instrumented binary, i.e., `start(__llvm_prf_cnts) - start(__llvm_prf_data)`.
```
done.

minglotus-6 commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+  and counter section in the instrumented binary, i.e., `start(__llvm_prf_cnts) - start(__llvm_prf_data)`.
+
+  It's used jointly with the `CounterPtr`_ field to compute the counter offset
+  relative to `start(__llvm_prf_cnts)`. Check out calculation-of-counter-offset_
```
done.

minglotus-6 commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+  for a visualized explanation.
+
+  .. note::
+    Instrumentations might not load the `__llvm_prf_data` object file section
```
done, and slightly reworded the sentence.

minglotus-6 commented on llvm/docs/InstrProfileFormat.rst:
```diff
@@ -0,0 +1,479 @@
+
+``BitmapDelta``
+  This field records the in-memory address difference between the `profile metadata`_
+  and bitmap section in the instrumented binary, i.e., `start(__llvm_prf_bits) - start(__llvm_prf_data)`.
```
done.

mingmingl-llvm committed:
resolve review feedback
